File: ruby/balancing_parentheses.rb
def balancing_parentheses(string)
  # type your code in here
  missing = 0
  openings = 0

  string.chars.each do |char|
    if char == '('
      openings+= 1
      next
    end

    if openings > 0
      openings -= 1
    else
      missing += 1
    end
  end

  missing + openings
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts balancing_parentheses('(()())')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses('()))')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses(')')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/balancing_parentheses.js
function balancingParentheses(string) {
  // type your code here
  let missing = 0
  let openings = 0

  for (let i = 0; i < string.length; i++) {
    if(string[i] === '(') {
      ++openings
      continue
    }

    if (openings > 0) {
      --openings
    } else {
      ++missing
    }
  }

  return missing + openings
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log(balancingParentheses('(()())'));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses('()))'));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses(')'));
}

module.exports = balancingParentheses;

// Please add your pseudocode to this file
// And a written explanation of your solution
